addons/account/models/account_move.py

```diff
@@ -2918,6 +2918,40 @@ def is_internal_partner(partner):
         move.message_subscribe(list(all_followers_ids))
         return move
 
+    def _post_check_invoice_non_negative(self):
+        """It checks that the total amount of a normal account cannot be less than zero.
+
+        In some accounting localizations this should be adjusted/disabled.
+        """
+        for move in self:
+            if (
+                move.is_invoice(include_receipts=True)
+                and float_compare(move.amount_total, 0.0, precision_rounding=move.currency_id.rounding) < 0
+            ):
+                raise UserError(_("You cannot validate an invoice with a negative total amount. "
+                                  "You should create a credit note instead. Use the action menu to transform it into "
+                                  "a credit note or refund."))
+
+    def _post_check_lock_date(self):
+        """When the accounting date is prior to a lock date, change it automatically upon posting.
+
+        /!\ 'check_move_validity' must be there since the dynamic lines will be recomputed outside the 'onchange'
+        environment.
+
+        In some accounting localizations this should be disabled.
+        """
+        for move in self:
+            affects_tax_report = move._affect_tax_report()
+            lock_dates = move._get_violated_lock_dates(move.date, affects_tax_report)
+            if lock_dates:
+                move.date = move._get_accounting_date(move.invoice_date or move.date, affects_tax_report)
+                move.with_context(check_move_validity=False)._onchange_currency()
```

Comment:
Not sure what kind of changes a l10n would have to this

Reply:
> `_post_check_lock_date`: eliminate this date modification, in hungary we should deny the post instead

I'm pretty sure that it is not a requirement. If you need to deny it, it must be because of other constraints that you can implement elsewhere (i.e `date == invoice_date`)

Reply:
> Not sure what kind of changes a l10n would have to this

The workflow:

* Someone create an incoming invoice, the accounting date is 2022-05-20
* some month passes (because of internal invoice validation processes), and the accountant locks the may month
* at 2022-07-20 the accountant see that there is a not posted invoice with accounting date 2022-05-20 (the actual lock date is 2022-05-31), so at the actual odoo code because of the lock date when she push the post button the accounting date will modify automatically to 2022-06-01.

But in hungary the accountant must decide that she want to activate this invoice at july (and not to june), so she needs to know that the original activation date is not usable anymore (because the Odoo lock date), so she can modify the activation date manually to anything she want.

I feel this must be implemented as a check *inside* the _post() method, and not as a `api.constraint`, because this check should be done during the posting of an invoice.

And another thing: that modification of the `move.date` *actually is inside* the actual _post() method, so how can i eliminate it with a `api.constraint`?

Reply:
> > `_post_check_lock_date`: eliminate this date modification, in hungary we should deny the post instead
> 
> I'm pretty sure that it is not a requirement. If you need to deny it, it must be because of other constraints that you can implement elsewhere (i.e `date == invoice_date`)

This is a valid customer requirement, what i cannot modify because it is *hardcoded* *inside* the actual _post() method!

Reply:
That is doable in `_get_accounting_date`